x-pack/plugins/alerting_builtins/server/alert_types/es_query/alert_type.ts

```diff
@@ -0,0 +1,134 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { i18n } from '@kbn/i18n';
+import { fold, fromNullable, mapNullable } from 'fp-ts/lib/Option';
+import { pipe } from 'fp-ts/lib/pipeable';
+import { find } from 'lodash';
+import { SavedObjectsErrorHelpers } from '../../../../../../src/core/server';
+import { AlertType, AlertExecutorOptions } from '../../types';
+import { ESQueryServices } from '.';
+import { Params } from './alert_type_params';
+import {
+  buildEsQuery,
+  getTime,
+  IIndexPattern,
+  IEsSearchResponse,
+} from '../../../../../../src/plugins/data/common';
+import { AlertServices } from '../../../../alerting/server';
+
+export const ID = '.es-query';
+
+const ActionGroupId = 'query matched';
+export function getAlertType(service: ESQueryServices): AlertType {
+  const alertTypeName = i18n.translate('xpack.alertingBuiltins.esQuery.alertTypeTitle', {
+    defaultMessage: 'ES Query',
+  });
+
+  const actionGroupName = i18n.translate(
+    'xpack.alertingBuiltins.esQuery.actionGroupQueryMatchedTitle',
+    {
+      defaultMessage: 'Query Matched',
+    }
+  );
+
+  return {
+    id: ID,
+    name: alertTypeName,
+    actionGroups: [{ id: ActionGroupId, name: actionGroupName }],
+    defaultActionGroupId: ActionGroupId,
+    executor,
+  };
+
+  async function executor({ params, services }: AlertExecutorOptions) {
+    const { query, filters = [], timefilter } = params as Params;
+    const index = find(filters, filter => !!filter.meta?.index)?.meta?.index;
+
+    const indexPattern = index
+      ? await findIndexPatternById(services.indexPattern, index)
+      : undefined;
+
+    const {
+      rawResponse: { hits: { total = 0, hits = [] } = {} },
+    } = (await services.search(
+      {
+        params: {
+          index,
+          body: {
+            query: buildEsQuery(
+              indexPattern,
+              query,
+              pipe(
+                fromNullable(timefilter),
+                mapNullable(timeFilter =>
+                  indexPattern ? getTime(indexPattern, timeFilter) : undefined
+                ),
+                fold(
+                  () => filters,
+                  timeFilter => [...filters, timeFilter]
+                )
+              )
```

Comment:
using `buildEsQuery` means that converting the SavedQuery to an ES DSL query has a one-to-one equivalence with what happens in Discover.
But it has a bunch of complications, such as needing the IndexPattern and potentially several type discrepancies that need to be ironed out.